### PR TITLE
Implement EIP-2537

### LINKED ISSUE #936: Implement EIP-2537
https://eips.ethereum.org/EIPS/eip-2537

### PR DESCRIPTION
**Pull request by gurukamath**
(closes #936 )

### What was wrong?
EIP-2537 needs is not currently implemented

Related to Issue #936 

### How was it fixed?
Implemented the EIP

#### Cute Animal Picture

Cute Animals - 1 of 1

### CODE REVIEW

**Review by petertdavies — CHANGES_REQUESTED** [review]
I've reviewed the `src/ethereum/prague/vm/precompiled_contracts/bls12_381/` and found a couple of issues, but the code generally looks good.

I'll do a further review when you rebase.

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+from py_ecc.optimized_bls12_381.optimized_curve import FQ as OPTIMIZED_FQ
+from py_ecc.optimized_bls12_381.optimized_curve import FQ2 as OPTIMIZED_FQ2
+from py_ecc.typing import Point2D
+
+from ethereum.base_types import U256, Bytes
+
+from ....vm.memory import buffer_read
+from ...exceptions import InvalidParameter
+
+P = FQ.field_modulus
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:34]: Updated

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    p = bytes_to_G1(buffer_read(data, U256(0), U256(128)))
+    if multiply(p, curve_order) is not None:
+        raise InvalidParameter("Sub-group check failed.")
+
+    m = int.from_bytes(buffer_read(data, U256(128), U256(32)), "big")
+
+    return p, m
+
+
+def bytes_to_FQ(
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:265]: Updated

**Comment by gurukamath** [comment]
Re-based on the latest `forks/prague`

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+    -------
+    point : Tuple[Point2D, int]
+        The G1 point and the scalar.
+
+    Raises
+    ------
+    InvalidParameter
+        If the sub-group check failed.
+    """
+    assert len(data) == 160
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:255]: Probably shouldn't be an assertion:

```suggestion
    if len(data) != 160:
        raise SomeException()
```

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+    -------
+    fq : Union[FQ, OPTIMIZED_FQ]
+        The FQ element.
+
+    Raises
+    ------
+    InvalidParameter
+        If the field element is invalid.
+    """
+    assert len(data) == 64
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:288]: Probably shouldn't be an assertion

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+    -------
+    fq2 : Union[FQ2, OPTIMIZED_FQ2]
+        The FQ2 element.
+
+    Raises
+    ------
+    InvalidParameter
+        If the field element is invalid.
+    """
+    assert len(data) == 128
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:323]: Same here.

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+    -------
+    point : Point2D
+        The G2 point.
+
+    Raises
+    ------
+    InvalidParameter
+        Either a field element is invalid or the point is not on the curve.
+    """
+    assert len(data) == 256
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:357]: Same

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+
+    x = bytes_to_FQ2(data[:128])
+    y = bytes_to_FQ2(data[128:])
+
+    assert isinstance(x, FQ2) and isinstance(y, FQ2)
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:362]: When wouldn't this be the case? Is `OPTIMIZED_FQ2` an `FQ2`?

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+    -------
+    point : Tuple[Point2D, int]
+        The G2 point and the scalar.
+
+    Raises
+    ------
+    InvalidParameter
+        If the sub-group check failed.
+    """
+    assert len(data) == 288
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:434]: Same assertion comment.

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/bls12_381_g1.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    K_DISCOUNT,
+    MAX_DISCOUNT,
+    MULTIPLIER,
+    G1_to_bytes,
+    bytes_to_FQ,
+    bytes_to_G1,
+    decode_G1_scalar_pair,
+)
+
+LEN_PER_PAIR = 160
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/precompiled_contracts/bls12_381/bls12_381_g1.py:35]:
```suggestion
LENGTH_PER_PAIR = 160
```

File: src/ethereum_spec_tools/evm_tools/t8n/__init__.py [hunk 10]
```diff
@@ -328,15 +328,6 @@ def apply_body(self) -> None:
...
 
-            self.fork.set_storage(
-                self.alloc.state,
-                self.HISTORY_STORAGE_ADDRESS,
-                (
-                    (self.env.block_number - 1) % self.HISTORY_SERVE_WINDOW
-                ).to_be_bytes32(),
-                U256.from_be_bytes(self.env.parent_hash),
-            )
-
```
Comment (SamWilsn) [inline src/ethereum_spec_tools/evm_tools/t8n/__init__.py:339]: This seems unrelated to the EIP in the PR title. Just want to make sure it is an intentional change.

File: src/ethereum_spec_tools/evm_tools/t8n/env.py [hunk 11]
```diff
@@ -238,12 +238,6 @@ def read_block_hashes(self, data: Any, t8n: "T8N") -> None:
         """
         Read the block hashes. Returns a maximum of 256 block hashes.
         """
-        # Read the block hashes
-        self.parent_hash = None
-        if t8n.fork.is_after_fork("ethereum.prague"):
-            self.parent_hash = Hash32(hex_to_bytes(data["parentHash"]))
-            return
-
```
Comment (SamWilsn) [inline src/ethereum_spec_tools/evm_tools/t8n/env.py:246]: Same here.

File: tests/prague/test_state_transition.py [hunk 12]
```diff
@@ -101,27 +101,35 @@ def test_general_state_tests(test_case: Dict) -> None:
...
+    "tests/fixtures/latest_fork_tests/fixtures/blockchain_tests/prague/eip2537_bls_12_381_precompiles/bls12_map_fp_to_g1",
+    "tests/fixtures/latest_fork_tests/fixtures/blockchain_tests/prague/eip2537_bls_12_381_precompiles/bls12_map_fp2_to_g2",
 )
-def test_execution_specs_generated_tests_7002(test_case: Dict) -> None:
-    run_prague_blockchain_st_tests(test_case)
 
 
-# Run execution-spec-generated-tests for EIP-6110
-test_dir = "tests/fixtures/latest_fork_tests/fixtures/blockchain_tests/prague/eip6110_deposits/deposits"
+def fetch_temporary_tests(test_dirs: Tuple[str, ...]) -> Generator:
```
Comment (SamWilsn) [inline tests/prague/test_state_transition.py:120]: Is this something we need to come back to later? Should we open an issue?

**Review by SamWilsn — COMMENTED** [review]
Just a general comment, but docstrings should be written in markdown now. Don't worry about fixing this PR, but going forward it'll make my life a bit easier.

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+    -------
+    point : Point2D
+        The G1 point.
+
+    Raises
+    ------
+    InvalidParameter
+        Either a field element is invalid or the point is not on the curve.
+    """
+    assert len(data) == 128
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:190]: These are utility functions. Calling them without the appropriate length of bytes is invalid behaviour. The assertion is a reminder for any function calling this utility in the future to pass the right size. As far as the current implmentation goes, the size checks are already performed in the calling function with the appropriate error raise. See `bls12_381_g1.bls12_g1_add` for example.

However, I would be ok with changing this to a `if...raise`

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py (hunk 7 above)
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:362]: `OPTIMIZED_FQ2` has been defined in `py_ecc` optimize mapping the field_element to curve. As such, the assertion was put in place to prevent `mypy` from complaining about the Union.

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+from py_ecc.typing import Point2D
+
+from ethereum.base_types import U256, Bytes
+
+from ....vm.memory import buffer_read
+from ...exceptions import InvalidParameter
+
+P = FQ.field_modulus
+
+K_DISCOUNT = {
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:36]: Updated

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/bls12_381_g1.py (hunk 9 above)
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/bls12_381_g1.py:35]: Updated

File: tests/prague/test_state_transition.py (hunk 12 above)
Comment (gurukamath) [reply tests/prague/test_state_transition.py:120]: This is to include only the tests which have the corresponding EIPs already implemented. Yes, this will go away as all the EIPs in the fork are implemented. This is already on the radar for post-prague re-factoring but I will create an issue.

File: src/ethereum_spec_tools/evm_tools/t8n/__init__.py (hunk 10 above)
Comment (gurukamath) [reply src/ethereum_spec_tools/evm_tools/t8n/__init__.py:339]: Good point. This is a remnant from the state of EIP-2935 and the state of the testing suite at the time the PR was raised. I will update this and check with EEST about whether we can continue to use the state tests for `evm_tools_tests`

File: src/ethereum/crypto/kzg.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+
+VERSIONED_HASH_VERSION_KZG = hex_to_bytes("0x01")
+BYTES_PER_COMMITMENT = 48
+BYTES_PER_PROOF = 48
+BYTES_PER_FIELD_ELEMENT = 32
+G1_POINT_AT_INFINITY = b"\xc0" + b"\x00" * 47
+BLS_MODULUS = 52435875175126190479447740508185965837690552500527637822603658699938581184513  # noqa: E501
+KZG_SETUP_G2_LENGTH = 65
+KZG_SETUP_G2_MONOMIAL_1 = "0xb5bfd7dd8cdeb128843bc287230af38926187075cbfbefa81009a2ce615ac53d2914e5870cb452d2afaaab24f3499f72185cbfee53492714734429b7b38608e23926c911cceceac9a36851477ba4c60b087041de621000edc98edada20c1def2"  # noqa: E501
```
[the remark this answers is not in the document]
Comment (petertdavies) [reply src/ethereum/crypto/kzg.py:38]: `BLS_MODULUS` is an integer.

`KZG_SETUP_G2_MONOMIAL_1` is an G2 elliptic curve point encoded in a standard binary format.

File: src/ethereum/crypto/kzg.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+
+class KZGCommitment(Bytes48):
+    """KZG commitment to a polynomial."""
+
+    pass
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/crypto/kzg.py:44]: Updated

File: src/ethereum/crypto/kzg.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (SamWilsn) [reply src/ethereum/crypto/kzg.py:38]: I guess my point is that `KZG_SETUP_G2_MONOMIAL_1` is conceptually not a string. It should probably be `Bytes`?

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (SamWilsn) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:190]: Yeah, I can accept that argument. I'll leave this up to your judgment.

My reasoning was that since `bytes_to_G1` accepts a `Bytes` as a parameter, it should handle every instance of that type. If this took, for example, a `Bytes128` it would make sense to use an assertion.

[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:190]: Sounds good. Updated

File: src/ethereum_spec_tools/evm_tools/t8n/__init__.py (hunk 10 above)
Comment (gurukamath) [reply src/ethereum_spec_tools/evm_tools/t8n/__init__.py]: Updated to the latest test suite with support for the `state-test` flag

File: src/ethereum_spec_tools/evm_tools/t8n/env.py (hunk 11 above)
Comment (gurukamath) [reply src/ethereum_spec_tools/evm_tools/t8n/env.py]: Updated

File: src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py (hunk 8 above)
Comment (gurukamath) [reply src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py:434]: Updated

**Review by SamWilsn — APPROVED** [review]

**Review by petertdavies — APPROVED** [review]
